BacktraceScreen.c

```diff
@@ -0,0 +1,384 @@
+
+   const BacktraceScreenDisplayOptions displayOptions = frame->backtracePanel->displayOptions;
+
+   char* functionName = frame->functionName;
+   int maxFunctionNameLength = printingHelper->maxFuncNameLen;
+   if (!!(displayOptions & DEMANGLE_NAME_FUNCTION) &&
+         printingHelper->maxDemangledFuncNameLen > 0) {
+      maxFunctionNameLength = printingHelper->maxDemangledFuncNameLen;
+      if (frame->demangleFunctionName) {
+         functionName = frame->demangleFunctionName;
+      }
+   }
+
+   char *completeFunctionName = NULL;
+   xAsprintf(&completeFunctionName, "%s+0x%lx", functionName, frame->offset);
+
+   char* objectDisplayed = frame->objectName;
+   size_t objectLength = printingHelper->maxObjNameLen;
+   if (!!(displayOptions & SHOW_FULL_PATH_OBJECT)) {
+      objectDisplayed = frame->objectPath;
+      objectLength = printingHelper->maxObjPathLen;
+   }
+
+   char* line = NULL;
+   int objectPathStart = -1;
+   int len = xAsprintf(&line, "%-*d 0x%0*lx %-*s %n%-*s",
+      (int)printingHelper->maxFrameNumLen, frame->index,
+      (int)printingHelper->maxAddrLen - (int)strlen("0x"), frame->address,
+      (int)maxFunctionNameLength, completeFunctionName,
+      &objectPathStart,
+      (int)objectLength, objectDisplayed ? objectDisplayed : "-"
+   );
+
+   int colors = CRT_colors[DEFAULT_COLOR];
+   if (!objectDisplayed) {
+      colors = CRT_colors[DYNAMIC_DARKGRAY];
+   }
+
+   RichString_appendnAscii(out, colors, line, len);
+
+   BacktraceFrame_highlightBasename(frame, out, line, objectPathStart);
+
+   free(completeFunctionName);
+   free(line);
+}
+
+static int BacktraceFrame_compare(const void* object1, const void* object2) {
+   const BacktraceFrame* frame1 = (const BacktraceFrame*)object1;
+   const BacktraceFrame* frame2 = (const BacktraceFrame*)object2;
+   return String_eq(frame1->functionName, frame2->functionName);
+}
+
+static void BacktraceFrame_delete(Object* object) {
+   BacktraceFrame* this = (BacktraceFrame*)object;
+   free(this->functionName);
+   free(this->demangleFunctionName);
+   free(this->objectPath);
+   free(this->objectName);
+   free(this);
+}
+
+static void BacktracePanel_setError(BacktracePanel* this, const char* error) {
+   assert(error);
+   assert(!this->error);
+
+   Panel* super = &this->super;
+   Panel_prune(super);
+
+   Vector* lines = this->super.items;
+   if (lines) {
+      Vector_delete(lines);
+   }
+
+   this->super.items = Vector_new(Class(ListItem), true, DEFAULT_SIZE);
+   Panel_set(super, 0, (Object*)ListItem_new(error, 0));
+}
+
+static void BacktracePanel_makePrintingHelper(const BacktracePanel* this, BacktracePanelPrintingHelper* printingHelper) {
+   Vector* lines = this->super.items;
+   size_t longestAddress = 0;
+
+   for (int i = 0; i < Vector_size(lines); i++) {
+      BacktraceFrame* frame = (BacktraceFrame*)Vector_get(lines, i);
+      frame->backtracePanel = this;
+
+      size_t digitOfOffsetFrame = strlen("+0x") + countDigit(frame->offset, 16);
```

Comment:
Now as I re-read this function, I think that maybe we don't need the `countDigit` function at all. You more likely need this instead: 

```c
size_t digitsOfOffsetFrame = (size_t)snprintf(NULL, 0, "+0x%zx", frame->offset);
```

And keep in mind you definition of `countDigit(0, 16)` would result in 0, but it should print one digit of "0" instead. So I believe you can convey your intent better with `snprintf`.

Reply:
I introduced `countDigit` because it's a utility function. I hope the developer will use it.
In terms of performance, `countDigit` should be a little bit better. And, with `countDigit` I found the code is easier to read.
So, I prefer that this function stays in the code.

However, you emphasize a beautiful bug 😄  Thank you!

Reply:
@MrCirdo
I don't think you should worry about performance right now, before profiling the code. As I mentioned in the last discussion, I doubt it is perfomant when you cast integer to float and then back to integer just for counting the digits.

Reply:
Or simply:

```c
size_t digitOfOffsetFrame = strlen("+0x") + countDigit(frame->offset | 1, 16);
```

And avoid the issue with `0x0` …

Calling `snprintf` here is like hunting flea with a cannon …

Reply:
> I don't think you should worry about performance right now, before profiling the code

Yes, you're alright. It was my intuition (and it could be wrong).

> Calling `snprintf` here is like hunting flea with a cannon …

Yes that's my concern. This is why I said it should be a little bit better in terms of performance.  

> And avoid the issue with 0x0 …

I fixed it by returning `1` when the digit is `0`.

So now, I prefer that the function stays in the code.

Reply:
@MrCirdo 
https://godbolt.org/z/8KGTfWdM1

```text
     100000000000000 -> 15 (expected 15)
     999999999999999 -> 16 (expected 15)
    1000000000000000 -> 16 (expected 16)
    9999999999999999 -> 17 (expected 16)
   10000000000000000 -> 17 (expected 17)
   99999999999999999 -> 18 (expected 17)
  100000000000000000 -> 18 (expected 18)
  999999999999999999 -> 19 (expected 18)
 1000000000000000000 -> 19 (expected 19)
 9999999999999999999 -> 20 (expected 19)
10000000000000000000 -> 20 (expected 20)
```

When I had an intuition that using floating point for `countDigit` isn't right here, this is what I got from the test.

And I can explain why this happened: The `double` type, when following IEEE 754 standard, only has 52 bits of significand and not 64. So large numbers like these can produce results that are rounded up.

Reply:
You point out something very interesting @Explorer09.

> And I can explain why this happened: The double type, when following IEEE 754 standard, only has 52 bits of significand and not 64. So large numbers like these can produce results that are rounded up.

I understand the problem, it's more clear for me.

I don't want to use `snprintf` function, It's too powerful.

I propose to revert `countDigit` with the logarithm function and put back `countDigit` with the counting loop.
@BenBE Is this right for you?